### PR TITLE
Restructure FunctionBody hierarchy for re-deferral

### PR DESCRIPTION
**Pull request by pleath**
Restructure the FunctionBody hierarchy so that FunctionInfo is a standalone proxy from which FunctionProxy does not inherit, and FunctionProxy is the basis for all the representations of user functions (FunctionBody, etc.). FunctionInfo still points to the FunctionProxy that implements the function, and FunctionProxy points to FunctionInfo. Do this to facilitate re-deferral and to maximize the memory benefit.

### CODE REVIEW

File: lib/Runtime/Language/FunctionCodeGenJitTimeData.h [hunk 1]
```diff
@@ -583,7 +583,7 @@ namespace Js
         bool IsPolymorphicCallSite(const ProfileId profiledCallSiteId) const;
         // This function walks all the chained jittimedata and returns the one which match the functionInfo.
         // This can return null, if the functionInfo doesn't match.
-        const FunctionCodeGenJitTimeData *GetJitTimeDataFromFunctionInfo(FunctionInfo *polyFunctionInfo) const;
+        const FunctionCodeGenJitTimeData *GetJitTimeDataFromFunctionInfo(FunctionInfo *polyFunctioInfoy) const;
```
Comment (ianwjhalliday) [inline lib/Runtime/Language/FunctionCodeGenJitTimeData.h]: nay i say, nay

File: lib/Backend/LowerMDShared.cpp [hunk 2]
```diff
@@ -1456,13 +1456,15 @@ LowererMD::Legalize(IR::Instr *const instr, bool fPostRegAlloc)
...
                 uint src1Forms = L_Reg | L_Mem | L_Ptr;     // Allow 64 bit values in x64 as well
-#if _M_X64
                 if (dst->IsMemoryOpnd())
                 {
+#if _M_X64
                     // Only allow <= 32 bit values
                     src1Forms = L_Reg | L_Imm32;
-                }
+#else
+                    src1Forms = L_Reg | L_Ptr;
```
Comment (ianwjhalliday) [inline lib/Backend/LowerMDShared.cpp]: Is this a related change?

File: lib/Runtime/Base/ScriptContext.cpp [hunk 3]
```diff
@@ -3189,6 +3189,8 @@ if (!sourceList)
         JavascriptMethod entryPoint = pFunction->GetEntryPoint();
         FunctionInfo * info = pFunction->GetFunctionInfo();
         FunctionProxy * proxy = info->GetFunctionProxy();
+
+#if 0
```
Comment (ianwjhalliday) [inline lib/Runtime/Base/ScriptContext.cpp]: is there a reason to keep this code around?

**Comment by ianwjhalliday** [comment]
Looks good to me other than a couple minor questions.

**Comment by digitalinfinity** [comment]
>         // WriteBarrier-TODO: Fix this? This is used only by proxies to keep the deserialized version around

Does this comment (around what this field is used for) need to be updated?

---
Refers to: lib/Runtime/Base/FunctionInfo.h:116 in b84eb2b. [](commit_id = b84eb2bf59a60a0647d0efcece9bc741d4855b53, deletion_comment = False)

File: lib/Runtime/Base/FunctionBody.h [hunk 4]
```diff
@@ -1348,16 +1393,180 @@ namespace Js
...
+    inline Js::LocalFunctionId FunctionProxy::GetLocalFunctionId() const
+    {
+        Assert(GetFunctionInfo());
+        return GetFunctionInfo()->GetLocalFunctionId();
+    }
+
+    inline void FunctionProxy::SetLocalFunctionId(LocalFunctionId functionId)
+    {
+        Assert(GetFunctionInfo());
+        Assert(GetFunctionInfo()->GetFunctionProxy() == this);
```
Comment (digitalinfinity) [inline lib/Runtime/Base/FunctionBody.h]: Can these asserts be moved to GetFunctionInfo itself? Also was curious why this particular assert wasn't there in GetLocalFunctionId (another reason I suggested centralizing the asserts)

Comment (pleath) [inline lib/Runtime/Base/FunctionBody.h]: There are cases where it’s awkward to assert in GetFunctionInfo() because we’re in the midst of swapping the functionBodyImpl. I suppose we could have a GetFunctionInfoUnchecked() for those.

--Paul

**Comment by dilijev** [comment]
@pleath any update on this PR?

**Comment by pleath** [comment]
Not today. I’ll be bringing it up to date soon.

-- Paul

**Comment by pleath** [comment]
Fully synched.

**Comment by dilijev** [comment]
Does https://github.com/Microsoft/ChakraCore/pull/1585 depend on this PR or are they independent changes?

**Comment by pleath** [comment]
Subsumed by https://github.com/Microsoft/ChakraCore/pull/1585.
